File: simpleMobileVMApp/simpleMobileVMAndroidLib/src/main/java/com/slickpath/mobile/android/simple/vm/machine/ProgramManager.java
package com.slickpath.mobile.android.simple.vm.machine;

import android.support.annotation.NonNull;

import com.slickpath.mobile.android.simple.vm.util.Command;

import java.util.ArrayList;
import java.util.List;

/**
 * This class abstracts the handling of the memory used to store program instructions and parameters
 *
 * @author Pete Procopio
 */
class ProgramManager {

    /**
     * Location in memory of stack pointer when stack is empty
     */
    static final int START_LOC = 0;

    /**
     * Value of memory if it is empty (unused)
     */
    private static final int EMPTY_MEMORY_VALUE = 999999;

    /**
     * Memory consists of N number of sequential Integers
     */
    private static final int MAX_MEMORY = 500;
    /**
     * Actually store we use for memory
     */
    private final List<Command> _programStore = new ArrayList<>(MAX_MEMORY);
    /**
     * Keeps track of location of next instruction to run
     */
    private int programCtr = START_LOC;
    /**
     * Keeps track of next location where an instruction to run can be written to- part of program setup
     */
    private int programWriter = START_LOC;

    /**
     * Constructor
     */
    ProgramManager() {
        // initialize every piece of instruction memory to EMPTY
        for (int i = 0; i < MAX_MEMORY; i++) {
            _programStore.add(new Command(EMPTY_MEMORY_VALUE, null));
        }
    }

    /**
     * Get current line in program that will execute next
     *
     * @return int
     */
    int getProgramCounter() {
        return programCtr;
    }

    /**
     * Sets value of program counter (set it to a position in memory where the next command will be run from)
     *
     * @param location location program counter points to
     */
    void setProgramCounter(final int location) {
        programCtr = location;
    }

    /**
     * Increment program counter location to next line of instruction
     */
    void incProgramCounter() {
        programCtr++;
    }

    /**
     * Decrement program counter location to previous line of instruction
     */
    void decProgramCounter() {
        programCtr--;
    }

    /**
     * program counter location to start of program memory
     */
    void resetProgramCounter() {
        programCtr = START_LOC;
    }

    /**
     * Get the current location in program memory where the next instruction will be added
     *
     * @return current location in program memory where the next instruction will be added
     */
    int getProgramWriterPtr() {
        return programWriter;
    }

    /**
     * Increment program writer location to previous memory location
     */
    void incProgramWriter() {
        programWriter++;
    }

    /**
     * Reset program writer location to starting memory location
     */
    void resetProgramWriter() {
        programWriter = START_LOC;
    }

    /**
     * Returns the instructions as a list
     * <p>
     * NOTE: This makes a copy of the program memory
     *
     * @return Copy of memory
     */
    @NonNull
    List<Command> dumpProgramStore() {
        return new ArrayList<>(_programStore);
    }


    /**
     * Returns the instruction at the specified location
     *
     * @param location location in program store
     * @return command at location
     */
    Command getCommandAt(final int location) {
        return _programStore.get(location);
    }

    /**
     * Set instruction value at the specified location
     *
     * @param location location in program store
     * @param command command that will be set
     */
    void setCommandAt(final int location, final Command command) {
        _programStore.set(location, command);
    }
}

File: simpleMobileVMApp/simpleMobileVMAndroidLib/src/main/java/com/slickpath/mobile/android/simple/vm/machine/ProgramManager.kt
package com.slickpath.mobile.android.simple.vm.machine

import com.slickpath.mobile.android.simple.vm.machine.MemoryStore.Companion.EMPTY_MEMORY_VALUE
import com.slickpath.mobile.android.simple.vm.machine.MemoryStore.Companion.MAX_MEMORY
import com.slickpath.mobile.android.simple.vm.util.Command

/**
 * Manages the program memory, storing instructions and providing access to them.
 *
 * This class handles the memory used to store program instructions and parameters. It manages
 * the program counter, program writer pointer, and provides methods to access and modify
 * the program store.
 *
 * @author Pete Procopio
 */
internal class ProgramManager {

    companion object {
        /**
         * The starting location of the program in memory (when the stack is considered empty).
         */
        const val START_LOCATION = 0
    }

    /**
     * The actual memory store for the program instructions.
     */
    private val programStore = MutableList(MAX_MEMORY) { Command(EMPTY_MEMORY_VALUE, null) }

    /**
     * The current instruction being pointed to.
     */
    private var programCounter = START_LOCATION

    /**
     * The next available location for writing an instruction.
     */
    var programWriterPointer = START_LOCATION
        private set

    /**
     * Sets the program counter to the specified location.
     *
     * @param location The new location for the program counter.
     * @throws IllegalArgumentException if the location is out of bounds.
     */
    fun setProgramCounter(location: Int) {
        require(location in 0 until MAX_MEMORY) { "Program counter location out of bounds: $location" }
        programCounter = location
    }

    /**
     * Gets the current location of the program counter.
     *
     * @return The current location of the program counter.
     */
    fun getProgramCounter(): Int {
        return programCounter
    }

    /**
     * Increments the program counter to the next instruction.
     *
     * @throws IllegalStateException if the program counter is at the end of memory.
     */
    fun incProgramCounter() {
        check(programCounter < MAX_MEMORY) { "Program counter cannot be incremented past the end of memory" }
        programCounter++
    }

    /**
     * Decrements the program counter to the previous instruction.
     *
     * @throws IllegalStateException if the program counter is at the beginning of memory.
     */
    fun decProgramCounter() {
        check(programCounter > 0) { "Program counter cannot be decremented past the beginning of memory" }
        programCounter--
    }

    /**
     * Resets the program counter to the start of the program.
     */
    fun resetProgramCounter() {
        programCounter = START_LOCATION
    }

    /**
     * Increments the program writer pointer to the next available location.
     *
     * @throws IllegalStateException if the program writer pointer is at the end of memory.
     */
    fun incrementProgramWriter() {
        programWriterPointer++
    }

    /**
     * Reset program writer location to starting memory location
     */
    fun resetProgramWriter() {
        check(programWriterPointer < MAX_MEMORY - 1) { "Program writer pointer cannot be incremented past the end of memory" }
        programWriterPointer = START_LOCATION
    }

    /**
     * Returns a copy of the program store.
     *
     * @return A copy of the program memory.
     */
    fun dumpProgramStore(): List<Command> {
        return programStore.toList()
    }

    /**
     * Returns the instruction at the specified location.
     *
     * @param location The location in the program store.
     * @return The command at the specified location.
     * @throws IllegalArgumentException if the location is out of bounds.
     */
    fun getCommandAt(location: Int): Command {
        return programStore[location]
    }

    /**
     * Set instruction value at the specified location
     *
     * @param location location in program store
     * @param command command that will be set
     */
    fun setCommandAt(location: Int, command: Command) {
        require(location in 0 until MAX_MEMORY) { "Set Location out of bounds: $location" }
        programStore[location] = command
    }

}